storage: check for conflicting replicated locks in mvccPutInternal

nvb opened the pull request:
Fixes #109647.
Informs #100193.

This commit uses the `lockTableKeyScanner` introduced in #110323 to scan for conflicting replicated locks at the same time as we can for intents (the writer's or others') when performing mutations in mvccPutInternal. In doing so, the commit removes use of `intentInterleavingIter` on the mutation path, opting for the explicit scan of the lock table instead.

Use of the `lockTableKeyScanner` to both check for conflicting locks and to scan for intents ensures that in the common case where there are no locks on a key, we only perform one (prefix) seek. This should avoid any regression. I'll collect some benchmark results to verify that this reasoning holds and that this does not cause a performance regression for write-heavy workloads.

Release note: None

nvb committed:
storage: check for conflicting replicated locks in mvccPutInternal

Fixes #109647.
Informs #100193.

This commit uses the `lockTableKeyScanner` introduced in #110323 to scan
for conflicting replicated locks at the same time as we can for intents
(the writer's or others') when performing mutations in mvccPutInternal.
In doing so, the commit removes use of `intentInterleavingIter` on the
mutation path, opting for the explicit scan of the lock table instead.

Use of the lockTableKeyScanner to both check for conflicting locks and
to scan for intents ensures that in the common case where there are no
locks on a key, we only perform one (prefix) seek. This should avoid any
regression. I'll collect some benchmark results to verify that this
reasoning holds and that this does not cause a performance regression
for write-heavy workloads.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110395)
<!-- Reviewable:end -->

nvb committed:
storage: break another dependency on intent interleaving in mvccPutInternal

This commit breaks another dependency on intent interleaving in
mvccPutInternal's call to replayTransactionalWrite. In this case,
we know the provisional value's timestamp, so we should use this
to read instead of the writer's read timestamp.

This was caught by the test:
```
TestReplicaTxnIdempotency/reissued_write_at_lower_timestamp
```

Release note: None

nvanbenschoten commented:
@arulajmani @sumeerbhola This PR is now rebased on master, so it's good for a look when possible.

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 10 of 10 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-:-NeLTGCI7aj7egPv3osY:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[`pkg/storage/mvcc.go` line 1645 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeL9CptEdXY8B7qBzf4:-NeL9CptEdXY8B7qBzf5:b-tltuj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/mvcc.go#L1645)):*
> ```Go
> 
> 	// TODO(nvanbenschoten): plumb in a maxConflicts value to limit the number
> 	// of shared locks that can be returned from the key.
> ```

Should we file an issue/add to our spreadsheet if we're not going to do this in this patch, just so we don't forget?
___
*[`pkg/storage/mvcc.go` line 1899 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeL9f7l83yl4cMyLp2v:-NeL9f7l83yl4cMyLp2w:bqweh6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/mvcc.go#L1899)):*
> ```Go
> 	putIsBlind := iter == nil
> 	if putIsBlind != (ltScanner == nil) {
> 		return false, errors.Errorf("iter and ltScanner must both be nil or both be non-nil")
> ```

nit: Should this be an assertion failed error?
___
*[`pkg/storage/mvcc.go` line 1936 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeLKmd--5sG5DPWH0Qq:-NeLKmd--5sG5DPWH0Qr:b-12uer5) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/mvcc.go#L1936)):*
> ```Go
> 				origRealKeyChanged = intentMeta.Timestamp.ToTimestamp()
> 			}
> 		}
> ```

I know it's implied by the comment inside of your if block, but should we add some words about the optimization to not look at the lock table if no version is found?
___
*[`pkg/storage/testdata/mvcc_histories/replicated_locks` line 357 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeLSfxt60LUl3hklV_X:-NeLSfxt60LUl3hklV_Y:b7gjsh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/testdata/mvcc_histories/replicated_locks#L357)):*
> ```
> 
> run error
> put t=A k=k1 v=v1
> ```

Is this the same as the case above? Was this meant to be t=B?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-:-NeLZHSHEhK6Qk0tIBgi:b-814g8k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/storage/mvcc.go` line 1645 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeL9CptEdXY8B7qBzf4:-NeLXK6jA826vfnwgBzb:bnv8xac) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/mvcc.go#L1645)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we file an issue/add to our spreadsheet if we're not going to do this in this patch, just so we don't forget?
</blockquote></details>

Yep, good point. Added to the spreadsheet.
___
*[`pkg/storage/mvcc.go` line 1899 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeL9f7l83yl4cMyLp2v:-NeLXW1P5WdiiX1c572i:b-pd8a1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/mvcc.go#L1899)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: Should this be an assertion failed error?
</blockquote></details>

I made the change here and for all of the other errors around it and found that it broke tests. For now, I added a TODO to switch them all over.
___
*[`pkg/storage/mvcc.go` line 1936 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeLKmd--5sG5DPWH0Qq:-NeLYjSk0tUSnyY-Pnil:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/mvcc.go#L1936)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I know it's implied by the comment inside of your if block, but should we add some words about the optimization to not look at the lock table if no version is found?
</blockquote></details>

Done.
___
*[`pkg/storage/testdata/mvcc_histories/replicated_locks` line 357 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-NeLSfxt60LUl3hklV_X:-NeLZ1gE8kbjoaG1_lOL:b383p8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/44344cf6a6ed3aabf51090668d6580610d6decb0/pkg/storage/testdata/mvcc_histories/replicated_locks#L357)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is this the same as the case above? Was this meant to be t=B?
</blockquote></details>

Oh yes, nice catch. Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 4 files at r4, 3 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-:-Nee2_ps46pKk00cvhHl:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/storage/mvcc.go` line 1692 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-Nedyu6c3Z_xlR2v6Nuu:-Nedyu6c3Z_xlR2v6Nuv:b-7dalqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd50c84ab632003e566b5fb67c62bf1524c3650c/pkg/storage/mvcc.go#L1692)):*
> ```Go
> // Since transactions should be idempotent, we must be particularly careful
> // about writing an intent if it was already written. If the sequence of the
> // transaction is at or below one found in `meta.Txn` then we should
> ```

Is this "If the sequence ..." slightly stale in that this seems to the only case in which this method is called?
___
*[`pkg/storage/mvcc.go` line 1647 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-Nedzq6p1iGKzK7SkGph:-Nedzq6p1iGKzK7SkGpi:bqbkv6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b84f5666fb61c71914431387b17e4b358ba0f9b/pkg/storage/mvcc.go#L1647)):*
> ```Go
> 
> 	// TODO(nvanbenschoten): plumb in a maxConflicts value to limit the number
> 	// of shared locks that can be returned from the key.
> ```

Is such a TODO also needed for the `MVCCPut` above?
___
*[`pkg/storage/mvcc.go` line 1915 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-Nee0knpEqy6QaM2DCO5:-Nee0knpEqy6QaM2DCO6:bwpz1fu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b84f5666fb61c71914431387b17e4b358ba0f9b/pkg/storage/mvcc.go#L1915)):*
> ```Go
> 		// necessary.
> 		ok, origMetaKeySize, origMetaValSize, origRealKeyChanged, err =
> 			mvccGetMetadata(iter, metaKey, &buf.meta)
> ```

`mvccGetMetadata` also has code to deal with an intent, which this call path will never encounter. Are we keeping that because there are other callers of `mvccGetMetadata` that need that?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-:-NefUn1O800IVEsqY3E6:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/storage/mvcc.go` line 1692 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-Nedyu6c3Z_xlR2v6Nuu:-NefTUPqAU_S-IE5-BAz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd50c84ab632003e566b5fb67c62bf1524c3650c/pkg/storage/mvcc.go#L1692)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this "If the sequence ..." slightly stale in that this seems to the only case in which this method is called?
</blockquote></details>

Done.
___
*[`pkg/storage/mvcc.go` line 1647 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-Nedzq6p1iGKzK7SkGph:-NefSy9t1YSvqLgtqnzy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b84f5666fb61c71914431387b17e4b358ba0f9b/pkg/storage/mvcc.go#L1647)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is such a TODO also needed for the `MVCCPut` above?
</blockquote></details>

Done.
___
*[`pkg/storage/mvcc.go` line 1915 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/110395#-Nee0knpEqy6QaM2DCO5:-NefTqfq-guF9VNf8xcz:bndwzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b84f5666fb61c71914431387b17e4b358ba0f9b/pkg/storage/mvcc.go#L1915)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

`mvccGetMetadata` also has code to deal with an intent, which this call path will never encounter. Are we keeping that because there are other callers of `mvccGetMetadata` that need that?
</blockquote></details>

Right, the function is still called by `MVCCGarbageCollect`. It still seems to work well for that caller and its use of an `intentInterleavingIter`, so I had no intention of changing it right now.


<!-- Sent from Reviewable.io -->
